File: lua/bank-account.lua
local BankAccount = {}

function BankAccount:new()
    local t = {}
    setmetatable(t, self)
    self.__index = self
    self._balance = 0
    return t
end

function BankAccount:balance()
    return self._balance
end

function BankAccount:deposit(amount)
    if self._isclose then
        return error("account already closed")
    end

    if amount <= 0 then
        return error("amount must > 0")
    end

    self._balance = self._balance + amount
end

function BankAccount:withdraw(amount)
    if self._isclose then
        return error("account already closed")
    end

    if amount <= 0 then
        return error("amount must > 0")
    end

    self._balance = self._balance - amount

    if self._balance < 0 then
        return error("balance must >= 0")
    end
end

function BankAccount:close()
    self._isclose = true
end

return BankAccount

File: python/bank_account.py
class BankAccount:
    def __init__(self):
        self._balance = 0
        self._isopen = False

    def get_balance(self):
        if not self._isopen:
            raise ValueError("account not open")
        return self._balance

    def open(self):
        if self._isopen:
            raise ValueError("account already open")
        self._isopen = True

    def deposit(self, amount):
        if not self._isopen:
            raise ValueError("account not open")
        if amount <= 0:
            raise ValueError("amount must be greater than 0")
        self._balance += amount

    def withdraw(self, amount):
        if not self._isopen:
            raise ValueError("account not open")
        if amount <= 0:
            raise ValueError("amount must be greater than 0")
        if amount > self._balance:
            raise ValueError("amount must be less than balance")
        self._balance -= amount

    def close(self):
        if not self._isopen:
            raise ValueError("account not open")
        self._isopen = False
        self._balance = 0
